Update tags on application definitions

douglasjacobsen opened the pull request:
This merge updates the tags used on application definitions, and their base classes. The goal is to minimize the number of unique tags used, to help simplify the use of tags by users.

dapomeroy commented on var/ramble/repos/builtin/applications/nvidia-hpl/application.py:
```diff
@@ -34,8 +34,6 @@ class NvidiaHpl(HplBase, NvidiaHPCBase):
 
     maintainers("douglasjacobsen")
 
-    tags("benchmark-app", "benchmark", "linpack", "optimized", "nvidia")
```
Should this include 'gpu-benchmark' or something that reflects the nvidia changes to base hpl?

Base HPL has the following tags:
        "hpc-benchmark",  # Key benchmark for TOP500 ranking
        "linpack",  # Based on the LINPACK benchmark
        "dense-linear-algebra",  # Solves a dense system of linear equations

douglasjacobsen replied:
Probably not here, but it could go on the NvidiaHPCBase class instead. That one has `gpu-accelerated` right now, but it could also have gpu-benchmark too.

dapomeroy reviewed (dismissed):
Overall LGTM.

I did a quick analysis of tag frequency and found that of 64 tags, 44% are used by more than one application (not counting inheritance from base applications):

Tag	Occurrences
cuda	6
hpc-benchmark	6
climate-modeling	4
memory-benchmark	4
molecular-dynamics	4
proxy-app	4
weather	4
cfd	3
communication-benchmark	3
gpu-benchmark	3
machine-learning	3
micro-benchmark	3
mpi-benchmark	3
network-benchmark	3
nvidia-tool	3
nwp	3
python	3
quantum-chemistry	3
system-utility	3
test-app	3
dft	2
diagnostics	2
finite-volume	2
intel-tool	2
io-benchmark	2
llm	2
materials-science	2
synthetic	2
aerospace	1
bioinformatics	1
build-tool	1
charmpp	1
circuit-simulation	1
conjugate-gradient	1
cosmology	1
dense-linear-algebra	1
dgemm	1
electronic-structure	1
euler	1
gpu-accelerated	1
gpu-utility	1
hidden-markov-models	1
hpc-ml	1
intel-optimized	1
jax	1
lapw	1
linpack	1
metadata-benchmark	1
mixed-precision	1
molecular-simulation	1
multiphysics	1
nlp	1
nvidia-framework	1
nvidia-optimized	1
ocean-modeling	1
plane-wave	1
pseudopotential	1
pytorch	1
shock-physics	1
software-stack	1
solver	1
spack-meta	1
storage-benchmark	1
tensorflow	1

linsword13 commented on var/ramble/repos/builtin/applications/spack-stack/application.py:
```diff
@@ -35,7 +35,11 @@ class SpackStack(ExecutableApplication):
 
-    tags("software", "configuration")
+    tags(
+        "spack-meta",  # Related to Spack package manager functionality
```
How about just calling it `spack`?

douglasjacobsen replied:
SG

linsword13 commented on var/ramble/repos/builtin/applications/su2/application.py:
```diff
@@ -19,7 +19,12 @@ class Su2(ExecutableApplication):
+    tags(
+        "cfd",  # Computational Fluid Dynamics suite, solves problems involving fluid flow
+        "multiphysics",  # Capable of handling various coupled physics
+        "finite-volume",  # Common numerical scheme used
```
I kinda feel this tag is too specific, and it only applies to certain solvers. It also has some solvers that use FEM: https://screenshot.googleplex.com/73BLhaejtxWsZoR.

douglasjacobsen replied:
I removed this one from all of the apps with it on them.

linsword13 commented on var/ramble/repos/builtin/base_applications/hpcg/base_application.py:
```diff
@@ -27,7 +27,11 @@ class Hpcg(ExecutableApplication):
-    tags("benchmark-app", "mini-app", "benchmark")
+    tags(
+        "hpc-benchmark",  # HPCG is a common HPC benchmark
+        "solver",  # Solves a sparse linear system
```
"solver" sounds a bit too generic?

douglasjacobsen replied:
Agreed. Removed! :)

linsword13 commented on var/ramble/repos/builtin/applications/hpcc/application.py:
```diff
@@ -27,7 +27,11 @@ class Hpcc(ExecutableApplication):
-    tags("benchmark-app", "mini-app", "benchmark", "DGEMM")
+    tags(
+        "hpc-benchmark",  # HPCC is a suite of several HPC benchmarks (HPL, STREAM, DGEMM, etc.)
+        "proxy-app",  # Represents a range of HPC kernel performance
```
I kinda prefer calling it "miniapp" over "proxy"

douglasjacobsen replied:
Done

linsword13 commented on var/ramble/repos/builtin/applications/intel-mpi-benchmarks/application.py:
```diff
@@ -24,11 +24,16 @@ class IntelMpiBenchmarks(ExecutableApplication):
     """
 
-    name = "IntelMpiBenchmarks"
+    name = "intel-mpi-benchmarks"
```
I like this change, but would it be better to do that in a separate PR? (I am happy with keeping it though.)

douglasjacobsen replied:
I had to make a few changes like this because the tags were all wrong when looking at them. So, I'd prefer to just keep it here since the other ones are more involved than this one.

But I think I agree in general, it would be nice to have name changes like this in a separate PR.

It's kind of interesting though that the name stuff never raised an issue with our tests. There is one example in this PR where we used `name` as a variable after defining the name for the application. This changed the name of the application to be a FOM name instead of what the app should actually have been named.

rfbgo commented on var/ramble/repos/builtin/applications/elk/application.py:
```diff
@@ -29,7 +29,12 @@ class Elk(ExecutableApplication):
 
-    tags("LAPW", "density-functional-theory")
+    tags(
+        "quantum-chemistry",  # Performs quantum chemical calculations
```
I think a lot of these comments are just the same thing again and not super useful. Can we make them less verbose somehow?

douglasjacobsen replied:
Do you just want all of the comments removed? Or are there specific comments you think are less useful?

I had them added as a way to try and convey to people what the tag should relate to, but they also should be somewhat self explanatory.

rfbgo replied:
Some are more useful than others. I'd be OK with all removed, or some more selective middle ground

douglasjacobsen replied:
I removed them all, if there are ones you want added back in, let me know.

douglasjacobsen committed:
Update tags on application definitions

This commit updates the tags used on application definitions, and their
base classes. The goal is to minimize the number of unique tags used, to
help simplify the use of tags by users.
